flink-tests/src/test/java/org/apache/flink/test/checkpointing/IgnoreInFlightDataITCase.java

```diff
@@ -107,17 +111,27 @@ public void testIgnoreInFlightDataDuringRecovery() throws Exception {
     }
 
     private static class SumFailSink implements SinkFunction<Integer>, CheckpointedFunction {
-        public static long result;
-        public static long resultBeforeFail;
+        private static long result;
+        private static long resultBeforeFail;
+
+        public SumFailSink() {
+            resultBeforeFail = result = 0;
+        }
 
         @Override
-        public void invoke(Integer value) throws Exception {
+        public void invoke(Integer value, Context context) throws Exception {
             result += value;
         }
 
         @Override
         public void snapshotState(FunctionSnapshotContext context) throws Exception {
-            resultBeforeFail = result;
+            if (resultBeforeFail == 0) {
+                resultBeforeFail = result;
+            }
+
+            // Allow to send data from the awaited map in this case we can be sure that in-flight
+            // data exists.
+            SlowMap.checkpointReachSinkLatch.trigger();
```

Comment:
I guess the failure reason was that no in-flight data was generated, and ignoring it didn't make any difference. Is that correct?

If yes, I'm afraid that this line still doesn't guarantee in-flight data: the snapshot can complete before `SlowMap` actually injects anything. Maybe wait here for some signal from `SlowMap`?

nit: I'd incapsulate `latch.trigger()` into a method with some meaningful name for readability

Reply:
the snapshot can not be completed before emitting data because at least two of subtasks waiting in the middle of processing and they cannot handle the checkpoint barrier before emitting the current value but these value can not be emitted before Sink don't start the checkpoint which guarantees that at least these two records will be in-flight data.

But it is a good question how to be sure that the checkpoint started when somebody is already waiting on this latch. Ideally, the checkpoint should be triggered manually when the latch is taken but I don't sure that it is possible to do it in ITCase(or I am wrong?).

Reply:
So the state is expected in channel between the active Map-3 and the Sink, right? 
But isn't it possible that right after receiving the barrier, sink will "unblock" Map-1 and -2; receive the barriers immediately, before processing any elements from the active Map-3 - thereby completing checkpoint without any channel state?

> But it is a good question how to be sure that the checkpoint started when somebody is already waiting on this latch. Ideally, the checkpoint should be triggered manually when the latch is taken but I don't sure that it is possible to do it in ITCase(or I am wrong?).


It's possible to trigger savepoint from the outside, but it doesn't capture inflight data apparently.
One option is to wait in `StreamOperator.prepareSnapshotPreBarrier` (but need to replace Map with an Operator).
Another one is to decline checkpoints in Map until it waited enough. E.g. by throwing an exception and causing a restart (I think throwing `CheckpointException` may not work from `Map`, and we'll probably need to handle multple checkpoints instead of just one)
Or maybe just re-run the test?:) As long as we can clearly distinguish between missing channel state (test issue) and other failures this should be fine I think.

Reply:
I actually expect the state between Source and Map-1, Map-2 or between Map-1, Map-2 and Sink. So I know that some data was emited by Source but weren't achieve the Sink because Map-1 and Map-2 are waiting. When Sink receives the barrier from Map-3 any other data which were sent to/from  Map-1 or Map-2 before the checkpoint barrier are in-flight data. As I understand, if Source sent at least one record to Map-1 or Map-2 before checkpoint was triggered it should work because this record should become in-flight data in some gate Source-Map or Map-Sink. But maybe I wrong somewhere.

I implemented restarting the task if there are not enough data for in-flight data guarantees. I don't know it will help or not but will see.

> How do we prevent/handle records spanning multiple buffers in this test? If they are possible and we drop them, then failures are inevitable. So we'll have to re-run the test in that case.

This problem is ignored right now. As I understand, it hardly possible here because the value is integer. But in general, this problem is possible but right now I don't have a good idea how to handle it correctly.

Reply:
I think you are right regarding which channel state will be captured and spanning records. 
Maybe put leave it as a comment in the code?